shared/src/commonMain/kotlin/at/asitplus/wallet/app/common/presentation/MdocPresenter.kt

```diff
@@ -0,0 +1,144 @@
+package at.asitplus.wallet.app.common.presentation
+
+import at.asitplus.wallet.lib.iso.DeviceResponse
+import com.android.identity.cbor.Bstr
+import com.android.identity.cbor.Cbor
+import com.android.identity.cbor.CborArray
+import com.android.identity.cbor.Tagged
+import com.android.identity.documenttype.DocumentTypeRepository
+import com.android.identity.mdoc.request.DeviceRequestParser
+import com.android.identity.mdoc.sessionencryption.SessionEncryption
+import com.android.identity.mdoc.transport.MdocTransport
+import com.android.identity.mdoc.transport.MdocTransportClosedException
+import com.android.identity.mdoc.util.toMdocRequest
+import com.android.identity.util.Constants
+import io.github.aakira.napier.Napier
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.first
+import ui.viewmodels.authentication.PresentationStateModel
+import ui.viewmodels.authentication.PresentationViewModel
+
+
+// Based on the identity-credential sample code
+// https://github.com/openwallet-foundation-labs/identity-credential/tree/main/samples/testapp
+class MdocPresenter(
+    private val stateModel: PresentationStateModel,
+    private val presentationViewModel: PresentationViewModel,
+    private val mechanism: MdocPresentmentMechanism
+) {
+    internal suspend fun present(
+        dismissible: MutableStateFlow<Boolean>,
+        numRequestsServed: MutableStateFlow<Int>,
+        credentialSelected: (DeviceResponse) -> Unit,
+    ) {
+        val transport = mechanism.transport
+        // Wait until state changes to CONNECTED, FAILED, or CLOSED
+        transport.state.first {
+            it == MdocTransport.State.CONNECTED ||
+                    it == MdocTransport.State.FAILED ||
+                    it == MdocTransport.State.CLOSED
+        }
+        if (transport.state.value != MdocTransport.State.CONNECTED) {
+            stateModel.setCompleted(Error("Expected state CONNECTED but found ${transport.state.value}"))
+            return
+        }
+
+        try {
+            var sessionEncryption: SessionEncryption? = null
+            var encodedSessionTranscript: ByteArray? = null
+            while (true) {
+                Napier.i("Waiting for message from reader...")
+                dismissible.value = true
+                val sessionData = transport.waitForMessage()
+                dismissible.value = false
+                if (sessionData.isEmpty()) {
+                    Napier.i("Received transport-specific session termination message from reader")
+                    stateModel.setCompleted()
+                    break
+                }
+
+                if (sessionEncryption == null) {
+                    val eReaderKey = SessionEncryption.getEReaderKey(sessionData)
+                    encodedSessionTranscript =
+                        Cbor.encode(
+                            CborArray.builder()
+                                .add(
```

Comment:
That's really hard to read with those line breaks ... please refactor to extension methods or but it into one line to make it readable